src/Gui/TextEdit.cpp

```diff
@@ -363,8 +364,10 @@ void TextEditor::keyPressEvent (QKeyEvent * e)
 {
     if ( e->key() == Qt::Key_Tab ) {
         ParameterGrp::handle hPrefGrp = getWindowParameter();
+        // if this is a python editor, then default to converting tabs to spaces
+        bool isPythonEditor = dynamic_cast<PythonEditor*>(this) != nullptr;

```

Comment:
A base class should not have a dependency to any of its sub-classes. Also this change adds an inconsistency because by default the Python console will use tabs while a Python editor uses spaces.

I recommend to add a new class that is derived from TextEditor and serves as base class for PythonConsole and PythonEditor. Then to this class the whole keyPressEvent() method can be moved.

Reply:
I have made that change, but since PythonConsole inherited from TextEdit instead of TextEditor, the required changes were rather extensive.  Further complicating matters, PythonConsole also inherited WindowParameter, which is also inherited by one of its new base classes.  Seems like a lot of changes just to change a simple default parameter.  Initially, I expected to just change one or two lines of code.  But it all seems to be working as expected.

Note the removal of attach() and detach() in PythonConsole.cpp.  I was getting warnings in the report view about observer already attached and observer already detached, and removing those 2 lines prevented the warning.  There might be further things that could also be removed.